core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -2000,6 +2023,147 @@ private[spark] class DAGScheduler(
+      // TODO: SPARK-33701: Instead of waiting for a constant amount of time for finalization
+      // TODO: for all the stages, adaptively tune timeout for merge finalization
+      try {
+        Futures.allAsList(results: _*).get(shuffleMergeResultsTimeoutSec, TimeUnit.SECONDS)
+      } catch {
+        case _: TimeoutException =>
+          logInfo(s"Timed out on waiting for merge results from all " +
+            s"$numMergers mergers for shuffle $shuffleId")
+          timedOut.set(true)
+          eventProcessLoop.post(ShuffleMergeFinalized(stage))
+      }
+    }
+  }
+
+  private def processShuffleMapStageCompletion(shuffleStage: ShuffleMapStage): Unit = {
+    markStageAsFinished(shuffleStage)
+    logInfo("looking for newly runnable stages")
+    logInfo("running: " + runningStages)
+    logInfo("waiting: " + waitingStages)
+    logInfo("failed: " + failedStages)
+
+    // This call to increment the epoch may not be strictly necessary, but it is retained
+    // for now in order to minimize the changes in behavior from an earlier version of the
+    // code. This existing behavior of always incrementing the epoch following any
+    // successful shuffle map stage completion may have benefits by causing unneeded
+    // cached map outputs to be cleaned up earlier on executors. In the future we can
+    // consider removing this call, but this will require some extra investigation.
+    // See https://github.com/apache/spark/pull/17955/files#r117385673 for more details.
+    mapOutputTracker.incrementEpoch()
+
+    clearCacheLocs()
+
+    if (!shuffleStage.isAvailable) {
+      // Some tasks had failed; let's resubmit this shuffleStage.
+      // TODO: Lower-level scheduler should also deal with this
+      logInfo("Resubmitting " + shuffleStage + " (" + shuffleStage.name +
+        ") because some of its tasks had failed: " +
+        shuffleStage.findMissingPartitions().mkString(", "))
+      submitStage(shuffleStage)
+    } else {
+      markMapStageJobsAsFinished(shuffleStage)
+      submitWaitingChildStages(shuffleStage)
+    }
+  }
+
+  private[scheduler] def handleRegisterMergeStatuses(
+      stage: ShuffleMapStage,
+      mergeStatuses: Seq[(Int, MergeStatus)]): Unit = {
+    // Register merge statuses if the stage is still running and shuffle merge is not finalized yet.
+    if (runningStages.contains(stage) && !stage.shuffleDep.shuffleMergeFinalized) {
+      mapOutputTracker.registerMergeResults(stage.shuffleDep.shuffleId, mergeStatuses)
+    }
+  }
+
+  private[scheduler] def handleShuffleMergeFinalized(stage: ShuffleMapStage): Unit = {
+    // Only update MapOutputTracker metadata if the stage is still active. i.e not cancelled.
+    if (runningStages.contains(stage)) {
+      stage.shuffleDep.markShuffleMergeFinalized()
+      processShuffleMapStageCompletion(stage)
+    } else {
+      // TODO: SPARK-35549: Currently merge statuses results which come after shuffle merge
+      // TODO: is finalized is not registered.
```

Comment:
@venkata91 This comment still doesn't explain why we unregisterAllMergeResults  here when `ShuffleMergeFinalized` message is received if a stage is not running. 
As I commented earlier, I think this should happen when a stage is cancelled. If a stage cancellation is going to be handled with a different jira, that is SPARK-35536, then we should unregister all the merge results with that.

Reply:
Just adding more clarification. 
The below comment suggests that when the stage is cancelled then finalizeShuffleMerge will be cancelled and with that there is no guarantee that there is a ShuffleMergeFinalized event will be posted
```
 // TODO: SPARK-35536: Cancel finalizeShuffleMerge if the stage is cancelled
          // TODO: during shuffleMergeFinalizeWaitSec
``` 
So, unregistering of all the merge results should ideally happen when the stage is cancelled. 
Since that will be handled with another jira, I think it is better to handle the unregisteration of all merge results with that jira as well.

In case you still want to keep it here, then please add a comment saying something like: this is targeting a stage cancellation while the merge finalization is not completed. Also, add the explanation that how it will interfere with the new stage after cancellation and the TODO should point to `SPARK-35536` where we will reconsider if we need to unregister here.

Reply:
I think the TODO comment I have is in wrong place. I have fixed that and also commented on why we are unregistering all the merge results.

Stage cancellation issue on the client side for shuffle merge is handled as part of this JIRA. Stage cancellation issue on server side would be handled as part of SPARK-35536.

[SPARK-35536](https://issues.apache.org/jira/browse/SPARK-35536) is a different issue where the stage is cancelled during the `shuffleMergeFinalizeWaitSec` time where we are making requests to individual shuffle services to finalizeShuffleMerge. This is to cancel that request. Also we might need to request all the shuffle services to clean up the merged blocks.

[SPARK-35549](https://issues.apache.org/jira/browse/SPARK-35549) is trying to address all the other cases like the one mentioned there in the JIRA - `executor/node loss causing re-computation due to fetch failure and if the merge statuses gets registered very late then that can cause inconsistencies.` which is why it is safer to not register merge results which comes after shuffle merge is finalized.